## Problem Statement

### Business Context

The healthcare industry is rapidly evolving, with professionals facing increasing challenges in managing vast volumes of medical data while delivering accurate and timely diagnoses. The need for quick access to comprehensive, reliable, and up-to-date medical knowledge is critical for improving patient outcomes and ensuring informed decision-making in a fast-paced environment.

Healthcare professionals often encounter information overload, struggling to sift through extensive research and data to create accurate diagnoses and treatment plans. This challenge is amplified by the need for efficiency, particularly in emergencies, where time-sensitive decisions are vital. Furthermore, access to trusted, current medical information from renowned manuals and research papers is essential for maintaining high standards of care.

To address these challenges, healthcare centers can focus on integrating systems that streamline access to medical knowledge, provide tools to support quick decision-making, and enhance efficiency. Leveraging centralized knowledge platforms and ensuring healthcare providers have continuous access to reliable resources can significantly improve patient care and operational effectiveness.

**Common Questions to Answer**

**1. Diagnostic Assistance**: "What are the common symptoms and treatments for pulmonary embolism?"

**2. Drug Information**: "Can you provide the trade names of medications used for treating hypertension?"

**3. Treatment Plans**: "What are the first-line options and alternatives for managing rheumatoid arthritis?"

**4. Specialty Knowledge**: "What are the diagnostic steps for suspected endocrine disorders?"

**5. Critical Care Protocols**: "What is the protocol for managing sepsis in a critical care unit?"

### Objective

As an AI specialist, your task is to develop a RAG-based AI solution using renowned medical manuals to address healthcare challenges. The objective is to **understand** issues like information overload, **apply** AI techniques to streamline decision-making, **analyze** its impact on diagnostics and patient outcomes, **evaluate** its potential to standardize care practices, and **create** a functional prototype demonstrating its feasibility and effectiveness.

### Data Description

The **Merck Manuals** are medical references published by the American pharmaceutical company Merck & Co., that cover a wide range of medical topics, including disorders, tests, diagnoses, and drugs. The manuals have been published since 1899, when Merck & Co. was still a subsidiary of the German company Merck.

The manual is provided as a PDF with over 4,000 pages divided into 23 sections.

## Installing and Importing Necessary Libraries and Dependencies

In [ ]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.2.28 --force-reinstall --no-cache-dir -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.4/9.4 MB 136.5 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 225.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 262.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.6/44.6 kB 237.0 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
numba 0.60.0 requires numpy<2.1,>=1.22, but you have numpy 2.4.6 which is incompatible.


**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
# For installing the libraries & downloading models from HF Hub
!pip install huggingface_hub==0.35.3 pandas==2.2.2 tiktoken==0.12.0 pymupdf==1.26.5 langchain==0.3.27 langchain-community==0.3.31 chromadb==1.1.1 sentence-transformers==5.1.1 numpy==2.3.3 -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 4.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 52.0/52.0 kB 5.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.0/44.0 kB 4.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 564.3/564.3 kB 20.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 55.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 71.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 63.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 88.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.9/19.9 MB 31.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 486.6/486.6 kB 41.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 16.6/16.6 MB 90.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 278.2/278.2 kB 29.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.6/4

**Note**:
- After running the above cell, kindly restart the runtime (for Google Colab) or notebook kernel (for Jupyter Notebook), and run all cells sequentially from the next cell.
- On executing the above line of code, you might see a warning regarding package dependencies. This error message can be ignored as the above code ensures that all necessary libraries and their dependencies are maintained to successfully execute the code in ***this notebook***.

In [ ]:
import os
from google.colab import userdata

#Libraries for processing dataframes,text
import json,os
import tiktoken
import pandas as pd

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download


from llama_cpp import Llama


## Question Answering using LLM

#### Downloading and Loading the model

In [ ]:
from huggingface_hub import list_repo_files
os.environ["HF_TOKEN"] = "INPUT TOKEN HERE"


repo_id = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
file_name = "mistral-7b-instruct-v0.2.Q6_K.gguf"
model_path = hf_hub_download(
    repo_id=repo_id,
    filename=file_name
)

In [ ]:
llm = Llama(
    model_path=model_path,
    n_ctx=2300,
    n_gpu_layers=32,
    n_batch=512
)

AVX = 1 | AVX_VNNI = 0 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | SSSE3 = 1 | VSX = 0 | 


#### Response

In [ ]:
def response(query,max_tokens=512,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
print(response("What is the protocol for managing sepsis in a critical care unit?", temperature = 0.1))

Llama.generate: prefix-match hit




Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:

1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.
2. ABCs: Ensure airway patency, adequate breathing, and circulatory support. Provide high-flow oxygen via a non-rebreather mask or endotracheal tube if necessary. Initiate intravenous fluids to maintain adequate blood pressure and organ perfusion.
3. Antibiotics: Administer broad-spectrum antibiotics based on the suspected source of infection and local microbiology data. Consider obtaining cultures before administering antibiotics if feasible.
4. Source control: Id

Llama.generate: prefix-match hit


Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are general steps for managing sepsis in a critical care unit:

1. Early recognition: Recognize the signs and symptoms of sepsis early and initiate treatment as soon as possible. Sepsis can present with various clinical features, including fever or hypothermia, tachycardia or bradycardia, altered mental status, respiratory distress, and lactic acidosis.
2. ABCs: Ensure airway patency, adequate breathing, and circulatory support. Provide high-flow oxygen via a non-rebreather mask or endotracheal tube if necessary. Initiate intravenous fluids to maintain adequate blood pressure and organ perfusion.
3. Antibiotics: Administer broad-spectrum antibiotics based on the suspected source of infection and local microbiology data. Consider obtaining cultures before administering antibiotics if feasible.
4. Source control: Identify and address the source of infection, such as drainage of an abscess or debridement of necrotic tissue.
5. Hemodynamic support: Provide vasopressors to maintain mean arterial pressure (MAP) greater than 65 mmHg if necessary. Consider using inotropic agents to improve cardiac output if needed.
6. Fluid resuscitation: Administer intravenous fluids to maintain adequate tissue perfusion and organ function. Use crystalloids initially, followed by colloids or blood products as needed.
7. Vasopressor dosing: Titrate vasopressors to maintain MAP greater than 65 mmHg while minimizing the risk of organ dysfunction.
8. Renal replacement therapy: Consider initiating renal replacement therapy (RRT) in patients with sepsis-induced acute kidney injury or those who are at high risk for developing it.
9. Glucose control: Maintain strict glucose control to prevent hyperglycemia, which can worsen inflammation and organ dysfunction. Aim for a blood glucose level between 80 and 150 mg/dL.
10. Corticosteroids: Consider using corticosteroids


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
print(response("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?", top_p=0.98))

Llama.generate: prefix-match hit




Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain is typically located in the lower right quadrant of the abdomen and may start as a mild discomfort that gradually worsens over time. The pain may be constant or intermittent and may be aggravated by movement, deep breathing, or coughing.
2. Loss of appetite: People with appendicitis may lose their appetite due to abdominal pain or nausea.
3. Nausea and vomiting: Vomiting is a common symptom of appendicitis, especially in the later stages of the condition.
4. Fever: A fever of 100.4°F (38°C) or higher may be present in people with appendicitis.
5. Constipation or diarrhea: Some people with appendicitis experience constipation, while others have diarrhea.
6. Rebound tenderness: When the doctor presses on the abdo

Llama.generate: prefix-match hit


Appendicitis is a medical condition characterized by inflammation of the appendix, a small tube-shaped organ located in the lower right side of the abdomen. The symptoms of appendicitis can vary from person to person, but some common signs include:

1. Abdominal pain: The pain is typically located in the lower right quadrant of the abdomen and may start as a mild discomfort that gradually worsens over time. The pain may be constant or intermittent and may be aggravated by movement, deep breathing, or coughing.
2. Loss of appetite: People with appendicitis may lose their appetite due to abdominal pain or nausea.
3. Nausea and vomiting: Vomiting is a common symptom of appendicitis, especially in the later stages of the condition.
4. Fever: A fever of 100.4°F (38°C) or higher may be present in people with appendicitis.
5. Constipation or diarrhea: Some people with appendicitis experience constipation, while others have diarrhea.
6. Rebound tenderness: When the doctor presses on the abdomen, there may be tenderness and pain that disappears when pressure is released (rebound tenderness).
7. Guarding: People with appendicitis may involuntarily contract their abdominal muscles to protect the area from pain (guarding).

Appendicitis cannot be cured via medicine alone. Antibiotics may be used to treat any accompanying bacterial infection, but they do not address the underlying inflammation or the presence of the appendix itself. Surgery is the only effective treatment for appendicitis. The most common surgical procedure used to remove the appendix is called an appendectomy. This procedure can be performed as an open surgery or laparoscopically, depending on the individual's condition and the surgeon's preference. During an appendectomy, the appendix is removed along with any infected tissue in the surrounding area. After the surgery, patients typically need to stay in the hospital for a few days to recover before returning home.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
print(response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?", top_k=20))

Llama.generate: prefix-match hit




Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, and eyelashes.

The exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.

There are several treatments that have been shown to be effective in addressing sudden patchy hair loss:

1. Corticosteroids: These are anti-inflammatory drugs that can help reduce inflammation and suppress the immune system's attack on the hair follicles. They can be applied topically or taken orally, depending on the severity of the condition.
2. Minoxidil: This is a medication that has been shown to promote hair growth in some people with alopecia areata. It works by increasing bloo

Llama.generate: prefix-match hit


Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles. It can result in round or oval bald patches on the scalp, but it can also occur on other parts of the body such as the beard area, eyebrows, and eyelashes.

The exact cause of alopecia areata is not known, but it's believed to be related to a problem with the immune system. Some possible triggers for this condition include stress, genetics, viral infections, and certain medications.

There are several treatments that have been shown to be effective in addressing sudden patchy hair loss:

1. Corticosteroids: These are anti-inflammatory drugs that can help reduce inflammation and suppress the immune system's attack on the hair follicles. They can be applied topically or taken orally, depending on the severity of the condition.
2. Minoxidil: This is a medication that has been shown to promote hair growth in some people with alopecia areata. It works by increasing blood flow to the affected area and stimulating hair follicles to produce new hairs.
3. DHT blockers: Dihydrotestosterone (DHT) is a hormone that can contribute to hair loss. Some people with alopecia areata have high levels of DHT in their scalp. Finasteride and dutasteride are medications that can block the production of DHT, which may help promote hair growth.
4. Herbal remedies: Certain herbs such as saw palmetto, nettle root, and biotin have been shown to promote hair growth and improve hair health. However, more research is needed to determine their effectiveness in treating alopecia areata specifically.
5. Dietary changes: Some people with alopecia areata have reported that changing their diet helped improve their hair loss. Eating a balanced diet rich in vitamins and minerals, particularly iron, zinc, and biotin, may help support healthy hair growth.
6. Stress management: Stress can trigger or worsen alopecia areata. Practicing stress-reducing techniques such as meditation, yoga, or deep breathing exercises may help reduce the severity of hair loss.
7. Hair transplantation

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
print(response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?", top_k=20, top_p = 0.98))

Llama.generate: prefix-match hit




A person who has sustained a physical injury to the brain, also known as a traumatic brain injury (TBI), may require various treatments depending on the severity and location of the injury. Here are some common treatments recommended for individuals with TBIs:

1. Emergency care: In case of a severe TBI, the first priority is to provide emergency care to ensure the person's airway is clear, breathing is stable, and circulation is adequate. This may involve intubation, oxygen therapy, and intravenous fluids.
2. Surgery: Depending on the location and severity of the injury, surgery may be necessary to remove hematomas (clots) or repair skull fractures.
3. Medications: Various medications may be prescribed to manage symptoms associated with TBIs, such as pain relievers for headaches, anti-seizure medications, and sedatives to help the person relax and rest.
4. Rehabilitation: Rehabilitation is an essential component of treatment for individuals with TBIs. This may include physical thera

Llama.generate: prefix-match hit


A person who has sustained a physical injury to the brain, also known as a traumatic brain injury (TBI), may require various treatments depending on the severity and location of the injury. Here are some common treatments recommended for individuals with TBIs:

1. Emergency care: In case of a severe TBI, the first priority is to provide emergency care to ensure the person's airway is clear, breathing is stable, and circulation is adequate. This may involve intubation, oxygen therapy, and intravenous fluids.
2. Surgery: Depending on the location and severity of the injury, surgery may be necessary to remove hematomas (clots) or repair skull fractures.
3. Medications: Various medications may be prescribed to manage symptoms associated with TBIs, such as pain relievers for headaches, anti-seizure medications, and sedatives to help the person relax and rest.
4. Rehabilitation: Rehabilitation is an essential component of treatment for individuals with TBIs. This may include physical therapy to help improve mobility and strength, occupational therapy to help individuals regain skills necessary for daily living activities, speech-language therapy to address communication difficulties, and cognitive rehabilitation to help individuals recover memory and attention skills.
5. Assistive devices: Depending on the extent of the injury, assistive devices such as wheelchairs, walkers, or braces may be necessary to help individuals regain mobility and independence.
6. Support groups: Joining a support group can provide emotional and social support for individuals with TBIs and their families. These groups offer opportunities to connect with others who have experienced similar injuries and share experiences, coping strategies, and resources.
7. Lifestyle modifications: Making lifestyle modifications, such as getting enough rest, eating a healthy diet, and avoiding alcohol and drugs, can help promote recovery and improve overall health and well-being for individuals with TBIs.
8. Vocational counseling and training: For individuals who have sustained a TBI that affects their ability to work, vocational counseling and training may be necessary to help them find suitable employment or adjust to new job requirements.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
print(response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"))

Llama.generate: prefix-match hit




First and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:

1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.
2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.
3. Immobilize the leg: Use a splint, sling, or other available materials to immobilize the leg and prevent movement. Be sure not to apply too much pressure on the injury site.
4. Provide pain relief: Offer over-the-counter pain medication, such as acetaminophen or ibuprofen, to help manage pain.
5. Seek medical attention: If the fracture is severe or if you suspect that there may be other injuries, call for emergency medical assistance.

Once you've ensured the person's safety and stability, consi

Llama.generate: prefix-match hit


First and foremost, if you suspect that someone has fractured their leg while hiking, it's essential to ensure their safety and prevent further injury. Here are some necessary precautions:

1. Keep the person calm and still: Encourage them to remain as still as possible to minimize pain and prevent worsening the injury.
2. Assess the situation: Check for any signs of shock, such as pale skin, rapid heartbeat, or shallow breathing. If you notice these symptoms, seek medical help immediately.
3. Immobilize the leg: Use a splint, sling, or other available materials to immobilize the leg and prevent movement. Be sure not to apply too much pressure on the injury site.
4. Provide pain relief: Offer over-the-counter pain medication, such as acetaminophen or ibuprofen, to help manage pain.
5. Seek medical attention: If the fracture is severe or if you suspect that there may be other injuries, call for emergency medical assistance.

Once you've ensured the person's safety and stability, consider the following steps for their care and recovery:

1. Transportation: Arrange for transportation to a hospital or medical facility as soon as possible. If it's not an emergency situation, try to make arrangements with friends or family members to help transport the person.
2. Medical evaluation: A healthcare professional will assess the extent of the fracture and determine the best course of treatment. This may include immobilization with a cast or brace, surgery, or other interventions.
3. Rehabilitation: After the initial injury has been treated, the person will likely need to undergo rehabilitation to regain strength and mobility in the affected leg. This may involve physical therapy exercises, stretching, and other interventions.
4. Follow-up care: Regular follow-up appointments with a healthcare professional are essential to ensure proper healing and prevent complications.
5. Recovery time: The recovery time for a fractured leg can vary depending on the severity of the injury and the individual's overall health. It's important to be patient and allow the body adequate time to heal.
6. Prevention: Encourage the person to take steps to prevent future injuries, such as wearing appropriate footwear, using trekking poles, and taking breaks during long hikes to rest

### Observations: Base LLM Responses
*   **Knowledge Source**: The model relies entirely on its pre-trained knowledge base.
*   **Structure**: While the answers are detailed, they can sometimes be overly broad or lack a consistent structure.
*   **Accuracy Risk**: Without specific grounding documents, the model runs a higher risk of hallucinating treatments or providing generic information that may not align with strict medical manual guidelines.

## Question Answering using LLM with Prompt Engineering

In [ ]:
# simple system message query string
system_message = """
You are a helpful medical assisstant that will provide healthcare information and medical knowledge to patients.
Answer in a friendly and clear manner, while keeping answers concise.

User Questions will begin with the token ###Question: .

If the answer cannot be found respond with "I don't know."


"""

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
print(response(system_message + "###Question: What is the protocol for managing sepsis in a critical care unit?", top_p = 0.98))

Llama.generate: prefix-match hit




Answer: The management of sepsis in a critical care unit involves early recognition and prompt intervention. Here are some general steps:

1. Recognize the signs and symptoms of sepsis, which can include fever or hypothermia, rapid heart rate, rapid breathing, confusion, and low blood pressure.
2. Obtain blood cultures before starting antibiotics to identify the causative organism.
3. Administer broad-spectrum antibiotics as soon as possible based on the patient's clinical presentation and suspected source of infection.
4. Provide fluid resuscitation to maintain adequate blood pressure and organ perfusion. This may involve administering crystalloid or colloid solutions.
5. Monitor and support vital functions, including oxygenation and ventilation, cardiac output, and renal function.
6. Administer corticosteroids in certain situations, such as septic shock or severe sepsis with persistent hypotension despite fluid resuscitation.
7. Provide source control for the infection, which may i

Llama.generate: prefix-match hit


Answer: The management of sepsis in a critical care unit involves early recognition and prompt intervention. Here are some general steps:

1. Recognize the signs and symptoms of sepsis, which can include fever or hypothermia, rapid heart rate, rapid breathing, confusion, and low blood pressure.
2. Obtain blood cultures before starting antibiotics to identify the causative organism.
3. Administer broad-spectrum antibiotics as soon as possible based on the patient's clinical presentation and suspected source of infection.
4. Provide fluid resuscitation to maintain adequate blood pressure and organ perfusion. This may involve administering crystalloid or colloid solutions.
5. Monitor and support vital functions, including oxygenation and ventilation, cardiac output, and renal function.
6. Administer corticosteroids in certain situations, such as septic shock or severe sepsis with persistent hypotension despite fluid resuscitation.
7. Provide source control for the infection, which may involve surgical intervention or other measures to eliminate the source of infection.
8. Monitor and manage complications, such as coagulopathy, acute respiratory distress syndrome (ARDS), and multiple organ dysfunction syndrome (MODS).
9. Provide supportive care, including nutritional support, pain management, and psychological support for the patient and their family.
10. Continuously reassess the patient's clinical status and adjust treatment plans as needed based on new information or changes in condition.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
print(response(system_message + "###Question: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?", temperature = 0.1))

Llama.generate: prefix-match hit




Appendicitis is a medical condition characterized by inflammation of the appendix. Common symptoms include:
1. Sudden pain in the lower right abdomen that may start around the navel and then shifts to the right side.
2. Loss of appetite.
3. Nausea and vomiting.
4. Fever, which may be low-grade at first but can rise as high as 103 degrees Fahrenheit (39.4 degrees Celsius).
5. Abdominal swelling.
6. Diarrhea or constipation.
7. Feeling sick or unwell in general.

Appendicitis cannot be cured with medicine alone, as the inflammation can lead to the appendix bursting and causing a potentially life-threatening infection called peritonitis. The standard treatment for appendicitis is an appendectomy, which is a surgical procedure to remove the appendix. This surgery is typically performed laparoscopically, using small incisions and a camera to guide the surgeon. Recovery time varies but usually takes about 1-2 weeks.


Llama.generate: prefix-match hit


Appendicitis is a medical condition characterized by inflammation of the appendix. Common symptoms include:
1. Sudden pain in the lower right abdomen that may start around the navel and then shifts to the right side.
2. Loss of appetite.
3. Nausea and vomiting.
4. Fever, which may be low-grade at first but can rise as high as 103 degrees Fahrenheit (39.4 degrees Celsius).
5. Abdominal swelling.
6. Diarrhea or constipation.
7. Feeling sick or unwell in general.

Appendicitis cannot be cured with medicine alone, as the inflammation can lead to the appendix bursting and causing a potentially life-threatening infection called peritonitis. The standard treatment for appendicitis is an appendectomy, which is a surgical procedure to remove the appendix. This surgery is typically performed laparoscopically, using small incisions and a camera to guide the surgeon. Recovery time varies but usually takes about 1-2 weeks.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
print(response(system_message + "###Question: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?", top_k=20))

Llama.generate: prefix-match hit




There are several potential causes for sudden, patchy hair loss. Some common ones include:

1. Alopecia Areata: An autoimmune condition that causes hair loss in small patches on the scalp or other areas of the body. It often results in round bald spots.
2. Telogen Effluvium: This is a type of non-scarring hair loss where large numbers of hairs enter the resting phase (telogen) at once, causing shedding and thinning. Stress, hormonal changes, or certain medications can trigger this condition.
3. Traction Alopecia: Hair loss due to excessive pulling or tension on the hair, often from wearing tight hairstyles like braids or ponytails.
4. Nutritional deficiencies: Lack of essential nutrients like iron, zinc, or vitamin B12 can lead to hair loss.
5. Scalp infections: Fungal infections like ringworm or bacterial infections can cause patchy hair loss.

As for treatments, the best option depends on the underlying cause. For autoimmune conditions like alopecia areata, corticosteroids (topical

Llama.generate: prefix-match hit


There are several potential causes for sudden, patchy hair loss. Some common ones include:

1. Alopecia Areata: An autoimmune condition that causes hair loss in small patches on the scalp or other areas of the body. It often results in round bald spots.
2. Telogen Effluvium: This is a type of non-scarring hair loss where large numbers of hairs enter the resting phase (telogen) at once, causing shedding and thinning. Stress, hormonal changes, or certain medications can trigger this condition.
3. Traction Alopecia: Hair loss due to excessive pulling or tension on the hair, often from wearing tight hairstyles like braids or ponytails.
4. Nutritional deficiencies: Lack of essential nutrients like iron, zinc, or vitamin B12 can lead to hair loss.
5. Scalp infections: Fungal infections like ringworm or bacterial infections can cause patchy hair loss.

As for treatments, the best option depends on the underlying cause. For autoimmune conditions like alopecia areata, corticosteroids (topical or injected) may be prescribed to suppress the immune response and promote regrowth. Minoxidil, a medication applied directly to the scalp, can help stimulate hair growth in some cases of androgenetic alopecia and telogen effluvium. Nutritional supplements and dietary changes can address nutritional deficiencies. Scalp treatments like antifungal shampoos or antibiotics may be prescribed for scalp infections.

It's important to consult a healthcare professional for an accurate diagnosis and personalized treatment plan.

### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
print(response(system_message + "###Question: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?", temperature = 0.05, top_p=0.98))

Llama.generate: prefix-match hit




I'm glad you asked about brain injuries! The treatment for a brain injury can depend on the severity and location of the injury. For mild traumatic brain injuries, also known as concussions, rest, hydration, and avoiding activities that could worsen symptoms are usually recommended. More severe injuries may require hospitalization, surgery, or rehabilitation. Rehabilitation therapies such as physical therapy, occupational therapy, speech therapy, and cognitive rehabilitation can help improve function and reduce symptoms. Medications may also be prescribed to manage symptoms like pain, swelling, or seizures. It's important for individuals with brain injuries to receive proper medical care and follow their healthcare provider's recommendations for treatment and recovery.


lama.generate: prefix-match hit


I'm glad you asked about brain injuries! The treatment for a brain injury can depend on the severity and location of the injury. For mild traumatic brain injuries, also known as concussions, rest, hydration, and avoiding activities that could worsen symptoms are usually recommended. More severe injuries may require hospitalization, surgery, or rehabilitation. Rehabilitation therapies such as physical therapy, occupational therapy, speech therapy, and cognitive rehabilitation can help improve function and reduce symptoms. Medications may also be prescribed to manage symptoms like pain, swelling, or seizures. It's important for individuals with brain injuries to receive proper medical care and follow their healthcare provider's recommendations for treatment and recovery.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
print(response(system_message + "###Question: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"))

Llama.generate: prefix-match hit




A: I'm sorry to hear that you or someone else has sustained a leg fracture while hiking. Here are some general steps for first aid and care:

1. **Assess the situation**: Ensure the safety of both the injured person and any helpers. Check for signs of shock, such as pale skin, rapid heartbeat, or shallow breathing.

2. **Immobilize the leg**: Use a splint, sling, or other immobilizing device to prevent further damage and reduce pain. Be sure not to move the leg excessively.

3. **Control bleeding**: Apply direct pressure with a clean cloth to any wounds. Elevate the injured leg if possible to help reduce swelling.

4. **Seek medical attention**: It's essential to get professional medical care as soon as possible, especially for a fracture. A healthcare provider can assess the severity of the injury and provide appropriate treatment.

5. **Pain management**: Your healthcare provider may prescribe pain medication or suggest other methods like ice packs or heat therapy to help manage di

Llama.generate: prefix-match hit


A: I'm sorry to hear that you or someone else has sustained a leg fracture while hiking. Here are some general steps for first aid and care:

1. **Assess the situation**: Ensure the safety of both the injured person and any helpers. Check for signs of shock, such as pale skin, rapid heartbeat, or shallow breathing.

2. **Immobilize the leg**: Use a splint, sling, or other immobilizing device to prevent further damage and reduce pain. Be sure not to move the leg excessively.

3. **Control bleeding**: Apply direct pressure with a clean cloth to any wounds. Elevate the injured leg if possible to help reduce swelling.

4. **Seek medical attention**: It's essential to get professional medical care as soon as possible, especially for a fracture. A healthcare provider can assess the severity of the injury and provide appropriate treatment.

5. **Pain management**: Your healthcare provider may prescribe pain medication or suggest other methods like ice packs or heat therapy to help manage discomfort.

6. **Recovery process**: The recovery time for a leg fracture varies depending on the severity of the injury and the specific treatment required. Physical therapy may be necessary to regain strength, flexibility, and mobility in the affected leg.

7. **Follow-up care**: Regular follow-up appointments with your healthcare provider are crucial to monitor progress and ensure proper healing.

Remember that this information is meant to be a general guideline, and individual circumstances may require different approaches. Always consult a healthcare professional for personalized advice.

### Observations: LLM with Prompt Engineering
*   **Persona Alignment**: Adding the system prompt effectively shifts the tone. The model now acts more like a polite, structured medical assistant.
*   **Formatting**: Responses are generally more concise and user-friendly, directly addressing the user's queries in a clear manner.
*   **Limitation**: Despite the improved formatting, the model is still limited to its internal knowledge. It doesn't yet have access to the specific context required for precise, localized clinical protocols.

## Data Preparation for RAG

### Loading the Data

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

pdf_path = '/content/drive/MyDrive/Colab Notebooks/GreatLearningLLM/medical_diagnosis_manual.pdf'
pdf_loader = PyMuPDFLoader(pdf_path)
pdf = pdf_loader.load()

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Data Overview

#### Checking the first 5 pages

In [ ]:
for i in range(min(5, len(pdf))):
    print(f"--- Page {i+1} ---")
    print(pdf[i].page_content[:500])
    print("\n")

--- Page 1 ---
raymondshaw6420@yahoo.com
5DR0YB1HQ3
This file is meant for personal use by raymondshaw6420@yahoo.com only.
Sharing or publishing the contents in part or full is liable for legal action.


--- Page 2 ---
raymondshaw6420@yahoo.com
5DR0YB1HQ3
This file is meant for personal use by raymondshaw6420@yahoo.com only.
Sharing or publishing the contents in part or full is liable for legal action.


--- Page 3 ---
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    .................................


--- Page 4 ---
491
Chapter 44. Foot & Ankle Disorders    .................

#### Checking the number of pages

In [ ]:
print(f"Total number of pages: {len(pdf)}")

Total number of pages: 4114


### Data Chunking

In [ ]:
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
docs = text_splitter.split_documents(pdf)
print(f"Total number of chunks created: {len(docs)}")

Total number of chunks created: 18050


In [ ]:
docs[0].page_content[:500]

'raymondshaw6420@yahoo.com\n5DR0YB1HQ3\nThis file is meant for personal use by raymondshaw6420@yahoo.com only.\nSharing or publishing the contents in part or full is liable for legal action.'

In [ ]:
docs[-1].page_content[:500]

"Zollinger-Ellison syndrome 95, 199, 200-201, 910\nmastocytosis vs 1125\nMenetrier's disease vs 132\npeptic ulcer disease vs 134\nZolmitriptan 1721\nZolpidem 1709, 3103\nZonisamide 1701\nZoonotic diseases, cutaneous 718\nZoophobia 1498\nZoster (see Herpes zoster virus infection)\nZygomycosis 1332\nThe Merck Manual of Diagnosis & Therapy, 19th Edition\nZ\n4104\nraymondshaw6420@yahoo.com\n5DR0YB1HQ3\nThis file is meant for personal use by raymondshaw6420@yahoo.com only.\nSharing or publishing the contents in part o"

### Embedding

In [ ]:

from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings

embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")

/tmp/ipykernel_9265/2007565587.py:6: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

### Vector Database

In [ ]:
vectorstore = Chroma.from_documents(documents=docs, embedding=embedding_function, persist_directory="./chroma_db")

In [ ]:
vectorstore.embeddings

In [ ]:
vectorstore.similarity_search("Hypothermia",k=3)

### Retriever

In [ ]:
retriever = vectorstore.as_retriever(
    search_type='similarity',
    search_kwargs={"k": 3}
  )

In [ ]:
rel_docs = retriever.get_relevant_documents("How to relive pain from a fever?")
rel_docs

/tmp/ipykernel_9265/111477714.py:1: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  rel_docs = retriever.get_relevant_documents("How to relive pain from a fever?")


[Document(metadata={'total_pages': 4114, 'keywords': '', 'source': '/content/drive/MyDrive/Colab Notebooks/GreatLearningLLM/medical_diagnosis_manual.pdf', 'file_path': '/content/drive/MyDrive/Colab Notebooks/GreatLearningLLM/medical_diagnosis_manual.pdf', 'moddate': '2026-06-06T01:11:28+00:00', 'modDate': 'D:20260606011128Z', 'subject': '', 'page': 1296, 'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creationDate': 'D:20120615054440Z', 'creationdate': '2012-06-15T05:44:40+00:00', 'author': '', 'format': 'PDF 1.7', 'trapped': '', 'creator': 'Atop CHM to PDF Converter', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition'}, page_content="• Identifying any localizing symptoms (eg, headache, cough): These symptoms help narrow the range of\npossible causes. The localizing symptom may be part of the patient's chief complaint or identified only\nby specific questioning.\n• Determining whether the patient is seriously or chronically ill (particularly if such illness i

### System and User Prompt Template

In [ ]:
qna_system_message = """
You are a helpful medical assistant that will provide healthcare information and medical knowledge to patients based on the provided context.
If the answer cannot be found in the context, respond with "I don't know."
Answer in a friendly and clear manner, while keeping answers concise.
"""

qna_user_message_template = """
Context: {context}

###Question: {question}

Answer:
"""

### Response Function

In [ ]:
def generate_rag_response(user_input,k=3,max_tokens=512,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

## Question Answering using RAG

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
print(generate_rag_response("What is the protocol for managing sepsis in a critical care unit?"))

Llama.generate: prefix-match hit


The management of sepsis in a critical care unit involves several steps. First, supportive measures are taken to maintain the patient's vital signs and prevent further deterioration. This includes keeping the patient warm, controlling any bleeding, securing the airway and providing ventilatory support if necessary, and avoiding giving anything by mouth.

Next, treatment is initiated simultaneously with evaluation. Supplemental oxygen is provided, and if the patient is in shock or has inadequate ventilation, airway intubation with mechanical ventilation may be required. Two large IV catheters are inserted into separate peripheral veins, and a central venous line or intraosseous needle may be used if peripheral veins cannot be accessed promptly.

The patient is then given aggressive fluid resuscitation to restore tissue perfusion. Antibiotics are administered based on the suspected cause of sepsis. Surgical excision of infected or necrotic tissues and drainage of pus may also be necessar

Llama.generate: prefix-match hit

The management of sepsis in a critical care unit involves several steps. First, supportive measures are taken to maintain the patient's vital signs and prevent further deterioration. This includes keeping the patient warm, controlling any bleeding, securing the airway and providing ventilatory support if necessary, and avoiding giving anything by mouth.

Next, treatment is initiated simultaneously with evaluation. Supplemental oxygen is provided, and if the patient is in shock or has inadequate ventilation, airway intubation with mechanical ventilation may be required. Two large IV catheters are inserted into separate peripheral veins, and a central venous line or intraosseous needle may be used if peripheral veins cannot be accessed promptly.

The patient is then given aggressive fluid resuscitation to restore tissue perfusion. Antibiotics are administered based on the suspected cause of sepsis. Surgical excision of infected or necrotic tissues and drainage of pus may also be necessary. Supportive care, including maintaining adequate nutrition and preventing complications such as infection, stress ulcers, and gastritis, is provided throughout the patient's stay in the ICU.

In some cases, intensive control of blood glucose and administration of corticosteroids and activated protein C may be necessary to improve outcomes. The severity of sepsis can vary, so individualized treatment plans are important based on the patient's specific condition.

### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
print(generate_rag_response("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"))

Llama.generate: prefix-match hit


The common symptoms of appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. After a few hours, the pain shifts to the right lower quadrant. Pain increases with cough and motion. Classic signs are right lower quadrant direct and rebound tenderness located at McBurney's point. Additional signs include pain felt in the right lower quadrant with palpation of the left lower quadrant. Appendicitis cannot be cured via medicine, and surgical removal is the standard treatment. The procedure may involve draining an abscess if one has formed before the operation. Patients should receive IV antibiotics effective against intestinal flora and adequate fluid and electrolyte replacement.


Llama.generate: prefix-match hit


The common symptoms of appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia. After a few hours, the pain shifts to the right lower quadrant. Pain increases with cough and motion. Classic signs are right lower quadrant direct and rebound tenderness located at McBurney's point. Additional signs include pain felt in the right lower quadrant with palpation of the left lower quadrant. Appendicitis cannot be cured via medicine, and surgical removal is the standard treatment. The procedure may involve draining an abscess if one has formed before the operation. Patients should receive IV antibiotics effective against intestinal flora and adequate fluid and electrolyte replacement.

### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
print(generate_rag_response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"))

Llama.generate: prefix-match hit


Sudden patchy hair loss, also known as alopecia, can have various causes. The most common cause is androgenetic alopecia, which is an androgen-dependent hereditary disorder where dihydrotestosterone plays a major role. However, other possible causes include infections such as fungal or bacterial infections, drugs including chemotherapeutic agents, trichotillomania (a psychological condition characterized by compulsive hair pulling), or autoimmune disorders like lupus erythematosus, lichen planopilaris, and systemic sclerosis.

For treating alopecia, the first step is to identify and address any underlying causes. For instance, fungal or bacterial infections can be treated with topical or oral antifungals or antibiotics, respectively. Traction alopecia, which results from physical stress on the scalp, can be treated by eliminating the source of traction.

For non-scarring alopecia, minoxidil (2% for women and 2% or 5% for men) is an effective treatment option. Minoxidil prolongs the ana

Llama.generate: prefix-match hit


Sudden patchy hair loss, also known as alopecia, can have various causes. The most common cause is androgenetic alopecia, which is an androgen-dependent hereditary disorder where dihydrotestosterone plays a major role. However, other possible causes include infections such as fungal or bacterial infections, drugs including chemotherapeutic agents, trichotillomania (a psychological condition characterized by compulsive hair pulling), or autoimmune disorders like lupus erythematosus, lichen planopilaris, and systemic sclerosis.

For treating alopecia, the first step is to identify and address any underlying causes. For instance, fungal or bacterial infections can be treated with topical or oral antifungals or antibiotics, respectively. Traction alopecia, which results from physical stress on the scalp, can be treated by eliminating the source of traction.

For non-scarring alopecia, minoxidil (2% for women and 2% or 5% for men) is an effective treatment option. Minoxidil prolongs the anagen growth phase and gradually enlarges miniaturized follicles into mature terminal hairs. It should be applied topically to the affected area once or twice daily.

For autoimmune disorders, behavior modification techniques, clomipramine, or selective serotonin reuptake inhibitors (SSRIs) like fluoxetine, fluvoxamine, paroxetine, sertraline, and citalopram may be beneficial for treating trichotillomania. In severe cases of alopecia areata, corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA) may be considered.

It's important to note that not all treatments are effective for everyone, and some may have side effects. Consulting a healthcare professional is recommended for an accurate diagnosis and personalized treatment plan.


### Query 4:  What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
print(generate_rag_response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"))

Llama.generate: prefix-match hit


For a person with a physical injury to brain tissue causing temporary or permanent impairment of brain function, the focus is on supportive care and rehabilitation. Supportive care includes preventing systemic complications such as pneumonia, urinary tract infections, and thromboembolic disease. Providing good nutrition and preventing pressure ulcers are also important. For patients with stroke, physical and occupational therapy may modestly improve functioning. Early intervention by rehabilitation specialists is crucial for maximal functional recovery. This includes prevention of secondary disabilities, family education, and establishing baseline findings before starting therapy. The specific type and extent of rehabilitation therapy depend on the patient's abnormalities, which vary based on the location and severity of the injury.


Llama.generate: prefix-match hit


For a person with a physical injury to brain tissue causing temporary or permanent impairment of brain function, the focus is on supportive care and rehabilitation. Supportive care includes preventing systemic complications such as pneumonia, urinary tract infections, and thromboembolic disease. Providing good nutrition and preventing pressure ulcers are also important. For patients with stroke, physical and occupational therapy may modestly improve functioning. Early intervention by rehabilitation specialists is crucial for maximal functional recovery. This includes prevention of secondary disabilities, family education, and establishing baseline findings before starting therapy. The specific type and extent of rehabilitation therapy depend on the patient's abnormalities, which vary based on the location and severity of the injury.

### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
print(generate_rag_response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"))

Llama.generate: prefix-match hit


If you've fractured your leg while hiking, it's essential to follow these steps:
1. Immobilize the affected leg using a splint or brace to prevent any further damage. You can use items like sticks, clothing, or a makeshift sling for temporary support until professional help arrives.
2. Apply ice packs to reduce swelling and pain. Do not apply ice directly to the skin; instead, wrap it in a cloth.
3. Elevate your leg above heart level whenever possible to minimize swelling and improve blood flow.
4. Take over-the-counter pain relievers like ibuprofen or acetaminophen as prescribed by a healthcare professional to manage pain.
5. Seek medical attention as soon as possible. A fracture requires proper diagnosis and treatment from a healthcare provider. They may recommend immobilization using a cast, brace, or traction device.
6. Follow the recommended care instructions provided by your healthcare provider for wound care, if applicable. Keep the wound clean and dry to prevent infection.
7. O

Llama.generate: prefix-match hit


If you've fractured your leg while hiking, it's essential to follow these steps:
1. Immobilize the affected leg using a splint or brace to prevent any further damage. You can use items like sticks, clothing, or a makeshift sling for temporary support until professional help arrives.
2. Apply ice packs to reduce swelling and pain. Do not apply ice directly to the skin; instead, wrap it in a cloth.
3. Elevate your leg above heart level whenever possible to minimize swelling and improve blood flow.
4. Take over-the-counter pain relievers like ibuprofen or acetaminophen as prescribed by a healthcare professional to manage pain.
5. Seek medical attention as soon as possible. A fracture requires proper diagnosis and treatment from a healthcare provider. They may recommend immobilization using a cast, brace, or traction device.
6. Follow the recommended care instructions provided by your healthcare provider for wound care, if applicable. Keep the wound clean and dry to prevent infection.
7. Once the fracture has healed, begin gentle stretching exercises followed by strengthening exercises as advised by your healthcare provider to regain full mobility and function in your leg.
8. Be aware of potential complications such as blood loss, fat embolism, or infection. Monitor for signs like excessive swelling, pain that worsens, fever, or coolness in the affected limb. If you notice any concerning symptoms, seek immediate medical attention.

### Observations: RAG Responses
*   **Contextual Accuracy**: The answers are now highly specific and strictly grounded in the extracted chunks from the `medical_diagnosis_manual.pdf`.
*   **Reduced Hallucinations**: By relying on the retriever, the model stops guessing and provides clinical facts, specific medications, and exact protocols detailed in the reference manual.
*   **Effectiveness**: This demonstrates the core value of RAG in healthcare—combining an LLM's natural language generation with an authoritative, localized knowledge base to ensure safe and reliable answers.

### Fine-tuning

In [ ]:
# Parameter Fine-Tuning Demonstrations

# Query 1: Testing max_tokens on a concise question
print("--- Query 1: Token Limit Check ---")
user_input_1 = "What is the maximum daily dose of acetaminophen?"
print(f"Query: {user_input_1}")
print(generate_rag_response(user_input_1, max_tokens=100))
print("\n# Observation: Even if max_tokens is set to 100, the model might stop early if the answer is concise, especially with temperature=0.")

# Query 2: Increasing temperature and max_tokens for a detailed bulleted list
print("\n--- Query 2: Temperature & Max Tokens for Detail ---")
user_input_2 = "List three common symptoms of appendicitis in bullet points and explain each briefly."
print(f"Query: {user_input_2}")
print(generate_rag_response(user_input_2, temperature=0.2, max_tokens=500))
print("\n# Observation: Increasing max_tokens allows the model to generate a complete list. The slight bump in temperature (0.2) allows slightly more natural phrasing.")

# Query 3: Adjusting top_p, top_k, and retrieval k for specific examples
print("\n--- Query 3: Top-p, Top-k, and increased Context (k) ---")
user_input_3 = "Can you explain specific clinical examples where active rewarming is required for hypothermia?"
print(f"Query: {user_input_3}")
print(generate_rag_response(user_input_3, top_p=0.98, top_k=20, max_tokens=350, k=5))
print("\n# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.")

# Query 4: Adjusting top_p, top_k, and retrieval k for specific examples
print("\n--- Query 4: Top-p, Top-k, and increased Context (k) ---")
user_input_4 = "Can you explain specific clinical examples where active rewarming is required for hypothermia?"
print(f"Query: {user_input_4}")
print(generate_rag_response(user_input_4, top_p=0.98, top_k=10, max_tokens=350, k=5))
print("\n# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.")

# Query 5: Adjusting top_p, top_k, and retrieval k for specific examples
print("\n--- Query 5: Top-p, Top-k, and increased Context (k) ---")
user_input_5 = "Can you explain specific clinical examples where active rewarming is required for hypothermia?"
print(f"Query: {user_input_5}")
print(generate_rag_response(user_input_5, top_p=0.90, top_k=20, max_tokens=350, k=5))
print("\n# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.")

--- Query 1: Token Limit Check ---
Query: What is the maximum daily dose of acetaminophen?


Llama.generate: prefix-match hit


The context does not provide information about the maximum daily dose of acetaminophen. However, if a patient has taken an overdose of acetaminophen, N-acetylcysteine may be given as an antidote to prevent liver damage. It is most effective when given within 8 hours of acetaminophen ingestion.

# Observation: Even if max_tokens is set to 100, the model might stop early if the answer is concise, especially with temperature=0.

--- Query 2: Temperature & Max Tokens for Detail ---
Query: List three common symptoms of appendicitis in bullet points and explain each briefly.


Llama.generate: prefix-match hit


1. Epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia: This means that the initial symptom is a feeling of pain in the upper abdomen (epigastrium) or around the belly button (periumbilical), which may be accompanied by nausea, vomiting, and loss of appetite. This pain is thought to occur due to the irritation of the appendix causing inflammation and pressure on surrounding organs.
2. Shifting pain to the right lower quadrant: After a few hours, the pain shifts from the upper abdomen to the right lower quadrant. This is because the appendix is located in the right lower part of the abdomen, and as it becomes more inflamed and distended, the pressure causes the pain to move to this area.
3. Pain increases with cough and motion: The pain worsens when the patient coughs or moves, indicating that there is inflammation and irritation in the appendix and surrounding tissue. This can make it difficult for the patient to find a comfortable position and may increas

Llama.generate: prefix-match hit


Sure! Active rewarming is necessary when hypothermia is severe or accompanied by cardiovascular instability, hormone insufficiency, or caused by trauma, toxins, or other conditions. For instance, a patient who has been submerged in cold water for an extended period and exhibits signs of shock or cardiac arrest would require active rewarming using methods like forced-air warming systems or heated infusion and lavage. Similarly, individuals with hypothermia due to hormone insufficiencies such as hypoadrenalism or hypothyroidism may also benefit from active rewarming to restore normal body temperature and improve overall health.

# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.

--- Query 4: Top-p, Top-k, and increased Context (k) ---
Query: Can you explain specific clinical examples where active rewarm

Llama.generate: prefix-match hit


Sure! Active rewarming is necessary when hypothermia is severe or accompanied by cardiovascular instability, hormone insufficiency, or caused by trauma, toxins, or other conditions. For instance, a patient who has been submerged in cold water for an extended period and exhibits signs of shock or cardiac arrest would require active rewarming using methods like forced-air warming systems or heated infusion and lavage. Similarly, individuals with hypothermia due to hormone insufficiencies such as hypoadrenalism or hypothyroidism may also benefit from active rewarming to restore normal body temperature and improve overall health.

# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.

--- Query 5: Top-p, Top-k, and increased Context (k) ---
Query: Can you explain specific clinical examples where active rewarm

Llama.generate: prefix-match hit


Sure! Active rewarming is necessary when hypothermia is severe or accompanied by cardiovascular instability, hormone insufficiency, or caused by trauma, toxins, or other conditions. For instance, a patient who has been submerged in cold water for an extended period and exhibits signs of shock or cardiac arrest would require active rewarming using methods like forced-air warming systems or heated infusion and lavage. Similarly, individuals with hypothermia due to hormone insufficiencies such as hypoadrenalism or hypothyroidism may also benefit from active rewarming to restore normal body temperature and improve overall health.

# Observation: If the context doesn't have exact examples, the model should say it doesn't know based on our system prompt. By increasing 'k' to 5, we broaden the search to give it the best chance of finding specific clinical conditions.


The changing of top_p, top_k does not affect the response of the LLM greatly. However, increasing the max_tokens allows the LLM to generate a full response without being cutoff. A longer max_tokens does increase the runtime for a response.

## Output Evaluation

Let us now use the LLM-as-a-judge method to check the quality of the RAG system on two parameters - retrieval and generation. We illustrate this evaluation based on the answeres generated to the question from the previous section.

- We are using the same Mistral model for evaluation, so basically here the llm is rating itself on how well he has performed in the task.

In [ ]:
groundedness_rater_system_message = """
You are an evaluator. Your task is to rate the groundedness of the generated answer based on the provided context.
The context is provided between <context> and </context> tags. The question is provided between <question> and </question> tags. The answer is provided between <answer> and </answer> tags.
Provide a score from 1 to 5, where 1 means completely ungrounded and 5 means fully grounded in the context.
"""

In [ ]:
relevance_rater_system_message = """
You are an evaluator. Your task is to rate the relevance of the generated answer to the provided question.
The question is provided between <question> and </question> tags. The answer is provided between <answer> and </answer> tags.
Provide a score from 1 to 5, where 1 means completely irrelevant and 5 means highly relevant.
"""

In [ ]:
user_message_template = """
<context>
{context}
</context>
<question>
{question}
</question>
<answer>
{answer}
</answer>
"""

In [ ]:
def generate_ground_relevance_response(user_input,k=3,max_tokens=350,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

### Query 1: What is the protocol for managing sepsis in a critical care unit?

In [ ]:
ground_score, rel_score = generate_ground_relevance_response("What is the protocol for managing sepsis in a critical care unit?")
print(f"Groundedness Evaluation:\n{ground_score}\n\nRelevance Evaluation:\n{rel_score}")

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


Groundedness Evaluation:
 I would rate the answer as 5. The answer is fully grounded in the context provided, as it accurately summarizes the information given about the management of sepsis in a critical care unit, including supportive care, fluid resuscitation, antibiotics, surgical intervention, and intensive care measures when necessary.

Relevance Evaluation:
 I would rate the answer as a 5. The answer fully addresses the question by providing a detailed description of the protocol for managing sepsis in a critical care unit, including supportive care, fluid resuscitation, antibiotics, surgical intervention, and supportive care measures. The context provided in the question is also referenced to provide additional context and detail.


Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit

Groundedness Evaluation:
 I would rate the answer as 5. The answer is fully grounded in the context provided, as it accurately summarizes the information given about the management of sepsis in a critical care unit, including supportive care, fluid resuscitation, antibiotics, surgical intervention, and intensive care measures when necessary.

Relevance Evaluation:
 I would rate the answer as a 5. The answer fully addresses the question by providing a detailed description of the protocol for managing sepsis in a critical care unit, including supportive care, fluid resuscitation, antibiotics, surgical intervention, and supportive care measures. The context provided in the question is also referenced to provide additional context and detail.


### Query 2: What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?

In [ ]:
ground_score, rel_score = generate_ground_relevance_response("What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?")
print(f"Groundedness Evaluation:\n{ground_score}\n\nRelevance Evaluation:\n{rel_score}")

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


Groundedness Evaluation:
 I would rate the answer as a 5. The symptoms described in the question are explicitly mentioned in the context, and the answer accurately reflects that appendicitis cannot be cured with medication and requires surgical removal.

Relevance Evaluation:
 Score: 5

The answer directly addresses the question by providing a clear description of the common symptoms of appendicitis and stating that it cannot be cured with medicine, requiring surgical removal instead.


Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit

Groundedness Evaluation:
 I would rate the answer as a 5. The symptoms described in the question are explicitly mentioned in the context, and the answer accurately reflects that appendicitis cannot be cured with medication and requires surgical removal.

Relevance Evaluation:
 Score: 5

The answer directly addresses the question by providing a clear description of the common symptoms of appendicitis and stating that it cannot be cured with medicine, requiring surgical removal instead.


### Query 3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?

In [ ]:
ground_score, rel_score = generate_ground_relevance_response("What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?")
print(f"Groundedness Evaluation:\n{ground_score}\n\nRelevance Evaluation:\n{rel_score}")

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


Groundedness Evaluation:
 I would rate the answer as 5. The answer is fully grounded in the context provided as it accurately identifies the condition (alopecia areata), discusses possible causes, and suggests effective treatments based on the information given in the context.

Relevance Evaluation:
 I would rate the answer as highly relevant (5). The answer accurately identifies the condition being discussed as alopecia areata and provides several effective treatment options, including topical and oral corticosteroids, topical immunotherapy, intralesional corticosteroid injections, minoxidil, and hair transplantation. The answer also emphasizes the importance of identifying the underlying cause of the sudden patchy hair loss before considering any treatment options.


Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit

Groundedness Evaluation:
 I would rate the answer as 5. The answer is fully grounded in the context provided as it accurately identifies the condition (alopecia areata), discusses possible causes, and suggests effective treatments based on the information given in the context.

Relevance Evaluation:
 I would rate the answer as highly relevant (5). The answer accurately identifies the condition being discussed as alopecia areata and provides several effective treatment options, including topical and oral corticosteroids, topical immunotherapy, intralesional corticosteroid injections, minoxidil, and hair transplantation. The answer also emphasizes the importance of identifying the underlying cause of the sudden patchy hair loss before considering any treatment options.

### Query 4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

In [ ]:
ground_score, rel_score = generate_ground_relevance_response("What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?")
print(f"Groundedness Evaluation:\n{ground_score}\n\nRelevance Evaluation:\n{rel_score}")

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


Groundedness Evaluation:
 I would rate the answer as a 5. The context and question both discuss the topic of brain injuries and recommended treatments. The answer accurately summarizes the information provided in the context regarding supportive care and rehabilitation therapy for individuals with brain injuries.

Relevance Evaluation:
 I would rate the answer as a 5 for relevance. The question asked about treatments for a person with a physical injury to brain tissue resulting in impairment of brain function. The answer provided detailed information on supportive care and rehabilitation therapy, which are specifically mentioned in the text as recommended treatments for such individuals.


### Query 5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

In [ ]:
ground_score, rel_score = generate_ground_relevance_response("What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?")
print(f"Groundedness Evaluation:\n{ground_score}\n\nRelevance Evaluation:\n{rel_score}")

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit


Groundedness Evaluation:
 I would rate the answer as a 5 for groundedness in the context. The context discusses various injuries, including lacerations and fractures, and their treatment. The question specifically asks about the necessary precautions and treatment steps for someone who has fractured their leg. The answer covers all of the relevant points from the context regarding fracture care, including rest, ice, compression, pain relief, medical evaluation, wound care (if applicable), and rehabilitation. Additionally, the answer mentions potential complications, which is also discussed in the context.

Relevance Evaluation:
 Score: 5

The answer is highly relevant to the question as it covers all the necessary precautions and treatment steps for a person who has fractured their leg, including rest, ice and compression, pain relief, medical evaluation, wound care (if applicable), and rehabilitation. The answer also mentions potential complications and the importance of following a h

Llama.generate: prefix-match hit
Llama.generate: prefix-match hit
Llama.generate: prefix-match hit

Groundedness Evaluation:
 I would rate the answer as a 5 for groundedness in the context. The context discusses various injuries, including lacerations and fractures, and their treatment. The question specifically asks about the necessary precautions and treatment steps for someone who has fractured their leg. The answer covers all of the relevant points from the context regarding fracture care, including rest, ice, compression, pain relief, medical evaluation, wound care (if applicable), and rehabilitation. Additionally, the answer mentions potential complications, which is also discussed in the context.

Relevance Evaluation:
 Score: 5

The answer is highly relevant to the question as it covers all the necessary precautions and treatment steps for a person who has fractured their leg, including rest, ice and compression, pain relief, medical evaluation, wound care (if applicable), and rehabilitation. The answer also mentions potential complications and the importance of following a healthcare provider's guidance.


### Observations: Output Evaluation (LLM-as-a-Judge)
*   **High Performance**: The RAG pipeline consistently scores very high (typically 5/5) on both Groundedness and Relevance.
*   **Groundedness**: The LLM judge successfully verifies that the RAG generator did not invent external facts and strictly utilized the provided `<context>` tags.
*   **Relevance**: The high relevance scores confirm that the retrieval process (the top-3 chunks) successfully fetched the exact information required to answer the specific medical queries.

## Actionable Insights and Business Recommendations

### Actionable Insights

1. **High-Fidelity Knowledge Retrieval**: The transition from a base LLM to a RAG-based approach successfully eliminated hallucinations and generic advice. By anchoring responses specifically to the *Merck Manuals*, the system guarantees that medical information retrieved is authoritative, accurate, and safe for reference.
2. **Tonal and Structural Consistency**: Prompt engineering proved highly effective in establishing a reliable "medical assistant" persona. The system reliably communicates in a professional, clear, and empathetic tone, which is vital for user trust in a healthcare setting.
3. **Adaptability through Parameter Tuning**: By fine-tuning retrieval metrics (like $k$) and generation parameters (like `temperature` and `max_tokens`), the system can be adapted for different use cases—such as providing rapid, concise answers for emergency room queries or detailed, comprehensive bullet points for treatment planning.

### Business Recommendations

1. **Integrate into Clinical Workflows (EHR)**: Embed this RAG-based AI directly into Electronic Health Record (EHR) systems. This will allow healthcare professionals to query diagnostic steps, drug interactions, and treatment protocols in real-time, drastically reducing the time spent sifting through heavy manuals or external databases during patient consultations.
2. **Standardization of Patient Care**: Deploy this tool across different hospital departments and branches to ensure that every practitioner—from seasoned specialists to junior doctors—has instant access to the exact same gold-standard medical protocols, reducing variability in patient care.
3. **Enhance Medical Training**: Utilize the AI assistant as an interactive educational tool for medical students and residents. They can safely query complex case scenarios and instantly receive textbook-grounded medical guidance to accelerate their learning curve.
4. **Continuous Knowledge Updates**: Establish a pipeline to routinely ingest new editions of medical manuals, clinical trial results, and localized hospital protocols into the vector database. This ensures the AI remains completely up-to-date with cutting-edge medical science without the immense cost and time required to fine-tune or retrain the underlying Large Language Model.

<font size=6 color='blue'>Power Ahead</font>
___

In [ ]:
import json

notebook_path = "/content/Full_Code_NLP_RAG_Project_Notebook_ (1).ipynb"

# Read the notebook
with open(notebook_path, 'r', encoding='utf-8') as f:
    nb = json.load(f)

# Remove widget state metadata which causes the KeyError
if 'widgets' in nb.get('metadata', {}):
    del nb['metadata']['widgets']

# Save the cleaned notebook
with open(notebook_path, 'w', encoding='utf-8') as f:
    json.dump(nb, f)

# Convert to HTML
!jupyter nbconvert --to html "{notebook_path}"

[NbConvertApp] Converting notebook /content/Full_Code_NLP_RAG_Project_Notebook_ (1).ipynb to html
[NbConvertApp] Writing 499431 bytes to /content/Full_Code_NLP_RAG_Project_Notebook_ (1).html
